accepted/0000-workspace-dependency-sharing.md

```diff
@@ -0,0 +1,283 @@
+tree like this:
+
+```
+root
++-- node_modules
+|   +-- foo (link: ../packages/foo)
+|   +-- bar (link: ../packages/bar) (forgot to update dependency!)
+|   +-- ... all other workspaces
++-- packages
+    +-- foo (v2)
+    +-- bar
+    |   +-- node_modules
+    |       +-- foo (v1)
+    +-- other workspaces
+        +-- node_modules (empty, getting foo@2 from top level)
+```
+
+If the user does not note the error, and publishes all of their packages
+together, they may end up releasing software that is incompatible with
+itself, and could have been detected during development.
+
+### Unable to support divergent peer dependencies
+
+It may be valuable in some cases to define multiple packages that each have
+a peer dependency on a _different_ version of some common module.
+
+For example, imagine a plugin that is designed to work alongside React, and
+has adapters to implement its functionality in different versions of React.
+
+```json
+{
+  "name": "my-react-plugin",
+  "version": "1.2.3",
+  "description": "Works with React v15 through v17!",
+  "optionalDependencies": {
+    "my-react-plugin-15": "*",
+    "my-react-plugin-16": "*",
+    "my-react-plugin-17": "*"
+  }
+}
+```
+
+Then, each adapter has a `package.json` like this:
+
+```json
+{
+  "name": "my-react-plugin-15",
+  "version": "1.2.3",
+  "peerDependencies": {
+    "react": "15"
+  }
+}
+```
+
+When a user runs `npm install my-reacrt-plugin`, it'll try to install all
+three optional dependencies, but 2 of them will always fail by virtue of
+the conflicting `react` peer dependency, and they will get one that works
+with their system.
+
+Unfortunately, doing this in a naive workspace implementation, where all
+`peerDependencies` are deduplicated to the top level means that the project
+will necessarily always fail to install with an `ERESOLVE` error, because
+the workspace projects cannot all be installed at the same time in the same
+place.
+
+Making this more complicated, if the intent of using workspaces is to
+ensure that all the workspaces within the project _may_ be used together,
+then isolating all conflicting `peerDependencies` would silently
+cause problems!  So, while we need a way to support this, it should not be
+the default behavior.
+
+## Detailed Explanation
+
+The following expectations should not be silently violated without explicit
+user intent.
+
+1. A workspace package that has a dependency on another workspace package
+   within the same project, should have its dependency met by the
+   package within the workspace, and not by fetching the dependency from
+   the registry.  If the workspace package does not satisfy the dependency
+   specification from another workspace package, and has not been
+   explicitly marked for isolation, then a warning will be displayed.
+2. A workspace that has a peer dependency should share the same instance of
+   that peer dependency with all other workspace projects, unless
+   explicitly marked for isolation.
```

Comment:
```suggestion
2. A workspace that has a peer dependency should share the same instance of
   that peer dependency with all other workspace projects that declare that peer dependency, unless
   explicitly marked for isolation.
```

Reply:
Will be provided by `--mode=isolated`, I'd prefer to leave it out of this change.

Reply:
I'm confused; if they don't declare the peer dep, and it's not provided by the graph, then it shouldn't be available at all.

In this case I'm not talking about the "accidental" availability because a transitive dep included it; i'm saying that with two sibling projects A and B, A declaring a peer dep on something shouldn't suddenly make it available to B.

Reply:
In isolated mode, it won't (assuming it isn't also an explicit root dependency).  However, in hoisted mode, anything "shared" between workspaces is shared by virtue of being hoisted, rather than symlinked into place.  So, yes, a peer dep of `A` will be available to workspace `B` whether it lists it as a dependency or not.

The only way I can see to add that constraint would be to say that we put the peer dep somewhere other than `./node_modules` and symlink it into all the workspaces that peer depend on it.  That feels like we're then basically doing the same thing as isolated mode, but only halfway, and blurs the expectation that "in hoisted mode, shared deps are hoisted; in isolated mode, shared deps are symlinked".

Reply:
Right - that “hallway” thing is precisely what the rfc i plan to write does :-) it’s fine if the current hoisting mechanism causes this behavior, but i was hoping that this RFC would maybe *be* the RFC i need to write :-p